File: compiler/src/pipeline/call_conventions.py
from __future__ import annotations

"""Built-in function call definitions and call conventions.

This module contains all the built-in function definitions that were previously
embedded in processor_base.py, along with the call convention classes.
"""

import re
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING

# Import proper Type objects for returns
from compiler_types.proper_types import STRING

if TYPE_CHECKING:
    from compiler_types.proper_types import Type


def js_field_access(s: str) -> str:
    """Convert field access to JavaScript syntax"""
    if re.fullmatch(r'[a-zA-Z_$][a-zA-Z0-9_$]*', s):
        return f'.{s}'
    return f'["{s}"]'


# "call conventions" are our sane way to emit JavaScript from arguments
# provided by child nodes

@dataclass
class FieldCall:
    """receiver.field"""
    field: str
    demands: list[str] | None
    returns: str | None
    def compile(self, args: list[str]):
        try:
            maybe_assign = ""
            receiver = args[0]
            args = args[1:]
            if len(args) > 0:
                # TODO - assert one arg (need context for compile errors...)
                #  or just this i guess..? what even happens lmfao
                maybe_assign = f" = ({",".join(args)})"
            return f"({receiver}.{self.field}{maybe_assign})"
        except Exception as e:
            raise Exception(f"{repr(self)} failed to compile") from e

@dataclass
class PrototypeCall:
    """String.join.call(self, args...) type shit"""
    constructor: str
    fn: str
    demands: list[str]
    returns: str
    def compile(self, args: list[str]):
        return f"{self.constructor}.prototype.{self.fn}.call({", ".join(args)})"
    
@dataclass
class DirectCall:
    """just call it directly fn(args...)"""
    fn: str
    receiver: str | None
    demands: list[str] | None
    returns: str | None
    def compile(self, args: list[str]):
        receiver = ""
        if self.receiver:
            receiver = f"{self.receiver}."
        return f"{receiver}{self.fn}({", ".join(args)})"

@dataclass
class LocalAccessCall:
    """just returns the identifier"""
    fn: str
    demands: list['Type'] | None
    returns: 'Type' | None
    def compile(self, args: list[str]):
        if len(args) > 0:
            return f"({self.fn} = {",".join(args)})"
        return self.fn


@dataclass
class NaryOperatorCall:
    """Emits a direct JavaScript operator for n-ary application (e.g., +, &&, ||)
    Optionally wraps the entire expression (e.g., for '!')"""
    operator: str
    demands: list[str] | None
    returns: str | None
    is_async: bool = False
    wrapper: str | None = None # New: Optional wrapper for the entire expression
    def compile(self, args: list[str]):
        expr = f"({f' {self.operator} '.join(args)})"
        if self.wrapper:
            return f"{self.wrapper}{expr}"
        return expr

@dataclass
class ChainedComparisonCall:
    """Emits a chained JavaScript comparison (e.g., a < b && b < c)"""
    operator: str
    demands: list[str] | None
    returns: str | None
    is_async: bool = False
    def compile(self, args: list[str]):
        if len(args) < 2:
            return "true" # A single element trivially satisfies a chained comparison
        comparisons = []
        for i in range(len(args) - 1):
            comparisons.append(f"{args[i]} {self.operator} {args[i+1]}")
        return f"({' && '.join(comparisons)})"

@dataclass
class NewCall:
    """Emits a new expression (e.g., new Thing(a, b))"""
    constructor: str
    demands: list[str] | None
    returns: str | None
    def compile(self, args: list[str]):
        return f"(new {self.constructor}({', '.join(args)}))"

@dataclass
class IndexAccessCall:
    """obj[index] or obj[index] = value"""
    demands: list[str] | None
    returns: str | None

    def compile(self, args: list[str]):
        receiver, index, *value = args
        if value:
            return f'({receiver}[{index}] = {value[0]})'
        return f'{receiver}[{index}]'

@dataclass
class CallableInvokeCall:
    """fn(args...) - invokes a callable/function"""
    demands: list[str] | None
    returns: str | None

    def compile(self, args: list[str]):
        fn, *fn_args = args
        return f"{fn}({', '.join(fn_args)})"

File: compiler/src_ts/pipeline/call_conventions.ts
// pipeline/call_conventions.ts
// Built-in function call definitions and call conventions.

import { Macro_context } from "../core/macro_registry.ts";
import { Emission_item, PARENTHESIS } from "../utils/strutil.ts";
import { map_adjacent } from "../utils/utils.ts";
import { FORCE_SYNTAX_ERROR } from "./js_conversion.ts";

// Convert a field name to JS access syntax
export function jsFieldAccess(s: string): string {
  if (/^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(s)) {
    return `.${s}`;
  } else {
    return `["${s}"]`;
  }
}

// --- Call conventions ---------------------------------------------------------

export enum Async_mode {
  ASYNC,
  SYNC,
}

const call = <T>(fn: () => T): T => fn();
const is_not_null = <T>(x: T | null): x is T => x !== null;

export class FieldCall {
  readonly field: string;
  readonly async_mode: Async_mode;
  constructor(
    opts: {
      field: string;
      async_mode: Async_mode;
    }
  ) { 
    this.field = opts.field;
    this.async_mode = opts.async_mode;
  }

  compile(args: Emission_item[], _ctx: Macro_context): [Emission_item, Emission_item] {
    const [fexpr_type, field_expr] = this.formatField(this.field);
      const args_real = args.filter(is_not_null);
    const throw_undefined = fexpr_type === "index" ? 
      () => {
        const receiver = args_real[0];
        if (!receiver) {
          return `/* invalid FieldCall: \`invalid${field_expr}\` */`;
        }
        return `if (${receiver()}${field_expr} === undefined) { throw new Error("sparse object") }`;
      }: null;
    return [throw_undefined, () => {
      const receiver = args_real[0];
      const rest = args_real.slice(1);
      const [ maybeAssign, wrap ] = rest.length === 0
        ? ["", (s: string) => s]
        : [` = (${rest.map(call).join(",")})`, PARENTHESIS];

      if (!receiver) {
        return `/* invalid FieldCall: \`invalid${field_expr}${maybeAssign}\` */`;
      }

      return wrap(`${receiver()}${field_expr}${maybeAssign}`);
    }];
  }

  private formatField(name: string): ["index" | "property", string] {
    if (this.safeIdentifier(name) && !this.keyword(name)) {
      return ["property", `.${name}`];
    } else {
      return ["index", `[${this.quoteString(name)}]`];
    }
  }

  private safeIdentifier(s: string): boolean {
    return /^[A-Za-z_]\w*$/.test(s);
  }

  private keyword(s: string): boolean {
    const keywords = new Set([
      "alias", "begin", "break", "case", "class", "def", "do", "else", "elsif",
      "end", "ensure", "enum", "extend", "false", "for", "if", "in", "include",
      "module", "next", "nil", "not", "of", "out", "private", "protected",
      "require", "rescue", "return", "self", "super", "then", "true", "typeof",
      "unless", "until", "when", "while", "with", "yield",
    ]);
    return keywords.has(s);
  }

  private quoteString(s: string): string {
    // Emit a double-quoted string literal with minimal escaping.
    const escaped = s
      .replace(/\\/g, "\\\\")
      .replace(/"/g, '\\"');
    return `"${escaped}"`;
  }
}

export class PrototypeCall {
  readonly constructorName: string;
  readonly fn: string;
  readonly async_mode: Async_mode;

  constructor(
    opts: {
      constructor: string,
      fn: string,
      async_mode: Async_mode,
    }
  ) {
    this.constructorName = opts.constructor;
    this.fn = opts.fn;
    this.async_mode = opts.async_mode;
  }

  compile(args: Emission_item[]): [Emission_item, Emission_item] {
    return [null, () => `${this.constructorName}.prototype.${this.fn}.call(${args.filter(is_not_null).map(call).join(", ")})`];
  }
}

// fn(args...)  (or receiver.fn(args...) if receiver provided)
export class DirectCall {
  readonly fn: string;
  readonly receiver: string | null = null;
  readonly async_mode: Async_mode;

  constructor(
    opts: {
      fn: string,
      receiver?: string | null,
      async_mode: Async_mode,
    }
  ) {
    this.fn = opts.fn;
    this.receiver = opts.receiver ?? null;
    this.async_mode = opts.async_mode;
  }

  compile(args: Emission_item[]): [Emission_item, Emission_item] {
    const prefix = this.receiver ? `${this.receiver}.` : "";
    return [null, () => `${prefix}${this.fn}(${args.filter(is_not_null).map(call).join(", ")})`];
  }
}

// just returns / assigns the identifier
export class LocalAccessCall {
  readonly fn: string;
  readonly async_mode: Async_mode;
  constructor(
    opts: {
      fn: string,
      async_mode: Async_mode,
    }
  ) {
    this.fn = opts.fn;
    this.async_mode = opts.async_mode;
  }

  compile(args: Emission_item[]): [Emission_item, Emission_item] {
    if (args.length > 0) {
      return [null, () => `(${this.fn} = ${args.filter(is_not_null).map(call).join(",")})`];
    } else {
      return [null, () => this.fn];
    }
  }
}

// emits a direct JS operator for n-ary application (e.g., +, &&, ||)
// optionally wraps the whole expression (e.g., "!" prefix)
export class NaryOperatorCall {
  readonly operator: string;
  readonly is_async: boolean = false;
  readonly wrapper: string | null = null;
  readonly async_mode: Async_mode;
  constructor(
    opts: {
      operator: string,
      is_async?: boolean,
      wrapper?: string | null,
      async_mode: Async_mode,
    }
  ) {
    this.operator = opts.operator;
    this.is_async = opts.is_async ?? false;
    this.wrapper = opts.wrapper ?? null;
    this.async_mode = opts.async_mode;
  }

  compile(args: Emission_item[]): [Emission_item, Emission_item] {
    return [null, () => {
      const expr = `(${args.filter(is_not_null).map(call).join(` ${this.operator} `)})`;
      if (this.wrapper) {
        return `${this.wrapper}${expr}`;
      } else {
        return expr;
      }
    }]
  }
}

// emits chained comparisons: a < b && b < c
export class ChainedComparisonCall {
  readonly operator: string;
  readonly async_mode: Async_mode;
  constructor(
    opts: {
      operator: string,
      async_mode: Async_mode,
    }
  ) {
    this.operator = opts.operator;
    this.async_mode = opts.async_mode;
  }

  compile(args: Emission_item[]): [Emission_item, Emission_item] {
    const args_real = args.filter(is_not_null);

    if (args_real.length < 2) {
      return [null, () => "true"];
    }

    return [null, () => 
      map_adjacent(args_real, (a, b) => 
        `(${a()} ${this.operator} ${b()})`
      ).join(" && ")
    ];
  }
}

// new Constructor(a, b)
export class NewCall {
  readonly constructorName: string;
  readonly async_mode: Async_mode;

  // only used for classes we define ourselves
  implementation: string | null = null;

  constructor(
    opts: {
      constructor: string,
      async_mode: Async_mode,
    }
  ) {
    this.constructorName = opts.constructor;
    this.async_mode = opts.async_mode;
  }

  compile(args: Emission_item[]): [Emission_item, Emission_item] {
    return [null, () => `(new ${this.constructorName}(${args.filter(is_not_null).map(call).join(", ")}))`];
  }
}

// obj[index]   or   obj[index] = value
export class IndexAccessCall {
  readonly async_mode: Async_mode;
  constructor(
    opts: {
      async_mode: Async_mode,
    }
  ) {
    this.async_mode = opts.async_mode;
  }

  compile(args: Emission_item[]): [Emission_item, Emission_item] {
    const args_real = args.filter(is_not_null);

    const receiver = args_real[0];
    const index = args_real[1];
    const value = args_real[2];

    if (receiver === undefined || index === undefined) {
      return [null, () => `${FORCE_SYNTAX_ERROR} /* invalid IndexAccessCall: \`${receiver?.() ?? "invalid"}[${index?.() ?? "invalid"}] = ${value?.() ?? "invalid"}\` */`];
    }

    if (value !== undefined) {
      return [null, () => `(${receiver()}[${index()}] = ${value()})`];
    } else {
      return [() => `if (${receiver()}[${index()}] === undefined) { throw new Error("sparse array") }`, () => `${receiver()}[${index()}]`];
    }
  }
}

// fn(args...) where the first arg is the callable
export class CallableInvokeCall {
  readonly async_mode: Async_mode;
  constructor(
    opts: {
      async_mode: Async_mode,
    }
  ) {
    this.async_mode = opts.async_mode;
  }

  compile(args: Emission_item[]): [Emission_item, Emission_item] {
    const args_real = args.filter(is_not_null);
    const fn = args_real[0];
    const params = args_real.slice(1);
    if (fn === undefined) {
      return [null, () => `${FORCE_SYNTAX_ERROR} /* invalid CallableInvokeCall: \`invalid(${params.map(call).join(", ")})\` */`];
    }
    return [null, () => `${fn()}(${params.map(call).join(", ")})`];
  }
}

// Union of all call-convention classes
export type Call_convention =
  | FieldCall
  | PrototypeCall
  | DirectCall
  | LocalAccessCall
  | NaryOperatorCall
  | ChainedComparisonCall
  | NewCall
  | IndexAccessCall
  | CallableInvokeCall;
